frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,339 @@
+        waveColor: "#A5A5AA",
+      })
+    })
+
+    recordPlugin.on("record-progress", time => {
+      setRecordingTime(formatTime(time))
+    })
+
+    setWavesurfer(ws)
+    setRecordPlugin(recordPlugin)
+
+    return () => {
+      if (ws) ws.destroy()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [theme, uploadTheFile])
+
+  useEffect(() => {
+    initializeWaveSurfer()
+  }, [initializeWaveSurfer])
+
+  const onClickPlayPause = useCallback(() => {
+    if (wavesurfer) {
+      wavesurfer.playPause()
+      // This is because we want the time to be the duration of the audio when they stop recording,
+      // but once they start playing it, we want it to be the current time. So, once they start playing it
+      // we'll start keeping track of the playback time from that point onwards (until re-recording).
+      setShouldUpdatePlaybackTime(true)
+
+      // to get the pause button to show
+      forceRerender()
+    }
+  }, [wavesurfer])
+
+  const startRecording = useCallback(() => {
+    if (!recordPlugin || !activeAudioDeviceId || !wavesurfer) {
+      return
+    }
+
+    const deviceId = activeAudioDeviceId
+    if (isNullOrUndefined(deviceId)) {
+      return
+    }
+
+    wavesurfer.setOptions({
+      waveColor: theme.colors.primary,
+    })
+
+    recordPlugin.startRecording({ deviceId: activeAudioDeviceId }).then(() => {
+      // Update the record button to show the user that they can stop recording
+      forceRerender()
+    })
+  }, [activeAudioDeviceId, recordPlugin, theme, wavesurfer])
+
+  const stopRecording = useCallback(() => {
+    if (!recordPlugin) return
+
+    recordPlugin.stopRecording()
+  }, [recordPlugin])
+
+  const handleClear = useCallback(() => {
+    if (isNullOrUndefined(wavesurfer) || isNullOrUndefined(deleteFileUrl)) {
+      return
+    }
+    setRecordingUrl(null)
+    wavesurfer.empty()
+    uploadClient.deleteFile(deleteFileUrl)
+    setProgressTime(STARTING_TIME_STRING)
+    setRecordingTime(STARTING_TIME_STRING)
+    setDeleteFileUrl(null)
+    widgetMgr.setFileUploaderStateValue(
+      element,
+      {},
+      { fromUi: true },
+      undefined
+    )
+    setShouldUpdatePlaybackTime(false)
+    if (notNullOrUndefined(recordingUrl)) {
+      URL.revokeObjectURL(recordingUrl)
+    }
+  }, [
+    deleteFileUrl,
+    recordingUrl,
+    uploadClient,
+    wavesurfer,
+    element,
+    widgetMgr,
+  ])
+
+  // Note: these can't be memoized due to the reliance on calling .isRecording() & .isPlaying()
+  const isPlayingOrRecording = Boolean(
+    (recordPlugin && recordPlugin.isRecording()) ||
+      (wavesurfer && wavesurfer.isPlaying())
+  )
+
+  const showPlaceholder = Boolean(
+    !(recordPlugin && recordPlugin.isRecording()) &&
+      !recordingUrl &&
+      !hasNoMicPermissions
+  )
+
+  const showNoMicPermissionsOrPlaceholder = useMemo(
+    () => Boolean(hasNoMicPermissions || showPlaceholder),
+    [hasNoMicPermissions, showPlaceholder]
+  )
```

Comment:
**suggestion (non-blocking):** I think it'd be fine to not wrap this in a `useMemo`, firstly because it matches the above boolean values, but also more broadly speaking because a `useMemo` is probably more overhead than necessary for this simple boolean comparison

Reply:
fair enough